Cache the OpenAPI schema for kubectl server-side dry run

julianvmodesto opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Server-side dry-run in kubectl makes one call to get the OpenAPI schema for every server-side dry-run request.

Instead, get the OpenAPI schema once for all server-side dry-run requests for a kubectl command.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
https://github.com/kubernetes/kubernetes/issues/97940

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

julianvmodesto committed:
Cache the OpenAPI schema for server-side dry-run.

Currently, server-side dry-run fetches the OpenAPI schema for every
single object.

This change fetches the OpenAPI schema only once.

julianvmodesto commented:
/wg api-expression
/sig cli
/area kubectl

julianvmodesto commented:
@apelisse I did some refactoring of the DryRunVerifier and the cmdutil factory, but I haven't wired this together yet for any of the commands.

How does this look?

apelisse commented on staging/src/k8s.io/kubectl/pkg/util/openapi/openapi_getter.go:
```diff
@@ -19,47 +19,36 @@ package openapi
 // NewOpenAPIGetter returns an object to return OpenAPIDatas which reads
 // from a server, and then stores in memory for subsequent invocations
-func NewOpenAPIGetter(openAPIClient discovery.OpenAPISchemaInterface) Getter {
+func NewOpenAPIGetter(openAPIClient discovery.OpenAPISchemaInterface) *synchronizedOpenAPIGetter {
```
Shouldn't you return `discovery.OpenAPISchemaInterface`? Can you even return a private type?

julianvmodesto replied:
Oh yeah sounds like I shouldn't do this

julianvmodesto replied:
Fixed.

apelisse commented on staging/src/k8s.io/cli-runtime/pkg/resource/dry_run_verifier.go:
```diff
@@ -29,15 +29,15 @@ import (
 // server-side dry-run. Otherwise, an error is returned.
-func VerifyDryRun(gvk schema.GroupVersionKind, dynamicClient dynamic.Interface, discoveryClient discovery.DiscoveryInterface) error {
-	verifier := NewDryRunVerifier(dynamicClient, discoveryClient)
+func VerifyDryRun(gvk schema.GroupVersionKind, dynamicClient dynamic.Interface, openAPIGetter discovery.OpenAPISchemaInterface) error {
```
Couldn't we just get the schema here? I don't know of the top of my head, but don't we always get the schema before creating the VerifyDryRun, why even using the lazy mechanism?

julianvmodesto replied:
Ah yes we can do that. No reason why it's lazy

julianvmodesto replied:
I found why it's lazy: we don't want to always get the OpenAPI document for every command where we implement dry run. Tests break when we fix this, and we'd have to inject a discovery.OpenAPISchemaInterface properly for tests where we don't need the document

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go:
```diff
@@ -159,19 +162,31 @@ func (f *factoryImpl) Validator(validate bool) (validation.Schema, error) {
+	return f.openAPIGetter.resources, f.openAPIGetter.err
+}
+
+func (f *factoryImpl) OpenAPIGetter() (discovery.OpenAPISchemaInterface, error) {
```
Where is this called?

julianvmodesto replied:
Nowhere, but I was planning on using this for the dry-run verifier. Maybe not a good idea... I'll take a look around

julianvmodesto replied:
This is now used

apelisse commented on staging/src/k8s.io/cli-runtime/pkg/resource/dry_run_verifier.go:
```diff
@@ -29,16 +29,23 @@ import (
-	return &DryRunVerifier{
-		finder:        NewCRDFinder(CRDFromDynamic(dynamicClient)),
-		openAPIGetter: discoveryClient,
+func NewDryRunVerifier(dynamicClient dynamic.Interface, openAPIGetter discovery.OpenAPISchemaInterface) (*DryRunVerifier, error) {
```
Why is this not taking the `openapi_v2.Document` directly? I'm wondering if this is going to cause some double-download of the OpenAPI?

maknihamdi replied:
we should modify all of kubectl cmd that calls this function

apelisse replied:
Are there so many calls?

julianvmodesto replied:
Yeah there are a lot of calls. I'll update one, but I want to make sure this looks okay before we update all of them

maknihamdi replied:
I think if you want to cache the document here, you can still use discoveryClient discovery.DiscoveryInterface et continue to cache an openapi, to pass it to DryRunVerifier struct

maknihamdi replied:
but I think you have to choose between caching the document in the Verifier 'constructor' ans continue using DiscoveryInterface or modify the constructor of the verifier to accept an OpenAPISchemaInterface the cache the call to the api. As I see you are using both in the same time

julianvmodesto replied:
PTAL, each subcommand is updated and this change builds properly

maknihamdi commented on staging/src/k8s.io/cli-runtime/pkg/resource/dry_run_verifier.go:
```diff
@@ -29,16 +29,23 @@ import (
 // server-side dry-run. Otherwise, an error is returned.
-func VerifyDryRun(gvk schema.GroupVersionKind, dynamicClient dynamic.Interface, discoveryClient discovery.DiscoveryInterface) error {
-	verifier := NewDryRunVerifier(dynamicClient, discoveryClient)
+func VerifyDryRun(gvk schema.GroupVersionKind, dynamicClient dynamic.Interface, openAPIDoc *openapi_v2.Document) error {
```
I think this function is never called. All of kubectl cmd use NewDryRunVerifier and (DryRunVerifier).HasSupport method to verify dryrun

apelisse replied:
That's a good point. We should probably remove it if it's not used. I think it used to, but the call site must have been refactored, thanks!

julianvmodesto replied:
Removed

apelisse reviewed (commented):
I think that looks good. I'm fine with the constructor taking the `OpenAPISchemaInterface` if it avoids changing a bunch of call-sites.

alvaroaleman commented:
@julianvmodesto @apelisse appears I did another PR for the same issue as I didn't see this one: https://github.com/kubernetes/kubernetes/pull/99104

Can we merge one of the two, please? The current state of affairs makes server side dry run effectively unusable for a larger amount of manifests.

maknihamdi commented:
> @julianvmodesto @apelisse appears I did another PR for the same issue as I didn't see this one: #99104
It's not the same approach. With your PR @alvaroaleman you still have 2 calls to open api, one in the Complete method https://github.com/kubernetes/kubernetes/blob/84483a5aacba8e535c9e185e9b0c01a5248b5086/staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go#L474 and one in the verifier, I think @julianvmodesto try to have only one call

alvaroaleman commented:
Yeah, the same problem, not the same approach. This PR doesn't compile right now though, can we get that fixed?

maknihamdi commented:
I think we should merge your PR @alvaroaleman,https://github.com/kubernetes/kubernetes/pull/99104. it is a fix for the main problem,  btw, with my team we do the same think as your PR to have a quick win, we build a custom kubectl ans use it, and we still waiting for this PR. I think in the PR we should do some refactoring with all of kubectl commands by moving the code of openApi Getter https://github.com/kubernetes/kubernetes/blob/84483a5aacba8e535c9e185e9b0c01a5248b5086/staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go#L45 to the cli-runtime package to be used by the verifier and the kubectl cmd

maknihamdi commented:
to optimize kubectl command, we also use an other trick: we use a pool of goroutine to visit resources https://github.com/kubernetes/kubernetes/issues/98886
https://github.com/kubernetes/kubernetes/pull/98888

alvaroaleman commented:
FTR, I think this is the better approach as it saves yet another openapi schema fetch, but also the more involved one

alvaroaleman commented:
@julianvmodesto you need to remove all the `go.sum` changes from this PR, presumably they are caused by you using go 1.16 already, but this repo is currently still on 1.15

julianvmodesto commented:
> @julianvmodesto you need to remove all the `go.sum` changes from this PR, presumably they are caused by you using go 1.16 already, but this repo is currently still on 1.15

Oh thank you!

fedebongio commented:
/triage accepted

alvaroaleman commented:
/retest

alvaroaleman commented:
@julianvmodesto given that this doesn't pass tests and is a lot more involved than https://github.com/kubernetes/kubernetes/pull/99104, would you be ok with https://github.com/kubernetes/kubernetes/pull/99104 merging and then rabasing this on top of that? The rebasing work should be minimal

julianvmodesto commented:
/retest

These look like flakes / unrelated failures

alvaroaleman commented:
> These look like flakes / unrelated failures

Other PRs do not have the amount of consistent failures this one has

apelisse commented:
/retest

apelisse commented:
Many of these error don't look related, but the integration tests clearly seem related. I don't know how much time will save by forcing this to be rebased.

alvaroaleman commented:
@apelisse I think a rebase of this on top of https://github.com/kubernetes/kubernetes/pull/99104 should be trivial.

The problem is that today, `--dry-run=server` is effectively unusable if you have more than a single digit number of manifests because it takes forever (and that problem is solved by both PRs). My concern is that this might not make it in time for 1.21 and the problem remains unsolved.

apelisse commented:
> My concern is that this might not make it in time for 1.21 and the problem remains unsolved.

I don't share that concern yet. We can merge the other one if this one doesn't merge before EOW, would that work?

alvaroaleman commented:
> I don't share that concern yet. We can merge the other one if this one doesn't merge before EOW, would that work?

Yup, sounds perfect!

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go:
```diff
@@ -61,6 +62,8 @@ type Factory interface {
+	// OpenAPISchema returns the parsed openapi schema definition
 	OpenAPISchema() (openapi.Resources, error)
+	// OpenAPIGetter returns a getter for the openapi schema document
+	OpenAPIGetter() discovery.OpenAPISchemaInterface
```
So my only objection is why would I care about these 2 separately instead of just having a single method `OpenAPISchema` do the heavy lifting for me like it did before with just the caching added?

apelisse replied:
Because unfortunately we have two different types: `openapi.Resources` which is our own format for the openapi, and the actual openapi. Some operations require the former, some operations require the latter 😞

soltysh replied:
I spoke with Antoine, it seems that the disconnection between `openapi.Resource` and `openapi_v2.Document` is too big to be addressed at this time. We agree that we'll revisit that at a later time. This is good to go as soon as it gets green.

soltysh reviewed (commented):
I've put a hold on @alvaroaleman #99104, I think this PR is closer to what I'd prefer seeing modulo my comment about `Factory` interface

stevekuznetsov commented:
@apelisse the end of the week came and went, this PR has merge conflicts, can we please move forward with one of the two and get this bug fixed? `kubectl apply` for some nontrivial amount of manifests takes _half an hour_ to run without the fix.

apelisse commented:
> @apelisse the end of the week came and went, this PR has merge conflicts, can we please move forward with one of the two and get this bug fixed? kubectl apply for some nontrivial amount of manifests takes half an hour to run without the fix.

@soltysh, I said we would merge the other PR at the end of next week if this one wasn't merged, I'd rather not be caught lying 😂 WDYT?

julianvmodesto commented:
I can rebase this on top of the other PR.

I definitely broke something not obvious to me yet in this PR...

soltysh commented:
> @soltysh, I said we would merge the other PR at the end of last week if this one wasn't merged, I'd rather not be caught lying joy WDYT?

I spoke with @stevekuznetsov and with @julianvmodesto I'll get this merged today/tomorrow at latest. The other approach is rather short-term fix, this is my preferred approach too :smile:

julianvmodesto commented:
Tests are fixed. I'm not sure why/how this works, but it's important not to cache until `f.clientGetter.ToDiscoveryClient()` returns properly.

apelisse commented:
/lgtm

Please squash the wip commit!

Thanks :-)

soltysh commented:
> Please squash the wip commit!

There's a neat way to squash commits on merge label that with `tide/merge-method-squash`, just fyi. I learned that recently from @rikatz

soltysh reviewed (approved):
/lgtm
/approve

julianvmodesto commented:
Squashed and rebased. pull-kubernetes-dependencies and pull-kubernetes-verify should pass after https://github.com/kubernetes/test-infra/pull/21094 is merged

apelisse commented:
Some non-flaky failures :-(

rikatz commented:
@apelisse not only you: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-dependencies

Probably some fix coming to test-infra in the next few minutes? :D

julianvmodesto commented:
/retest

rikatz commented:
@julianvmodesto if it fails again, PTAL at https://groups.google.com/g/kubernetes-dev/c/WJmMR2IhrNM if this is impacted by go 1.16 migration :) (copied from @eddiezane comment on other PR)

soltysh reviewed (approved):
/lgtm

spiffxp commented:
/hold
https://prow.k8s.io/?repo=kubernetes%2Fkubernetes&type=batch has been failing consistently, I am taking a random guess that this may be a poison PR that is no longer getting along with go1.16

spiffxp commented:
/test pull-kubernetes-unit

soltysh commented:
Let's double check and will drop if it gets green again
/test pull-kubernetes-unit

soltysh commented:
/test pull-kubernetes-unit

alvaroaleman commented:
/hold cancel

Given that it passed all tests I think its fair to assume that it is not going to break batches :)

alvaroaleman commented:
thanks a lot for your work on this, @julianvmodesto !

spiffxp commented:
Thanks! And sorry for the random /hold, was in no way a comment on the quality of your contribution
